File: scripts/cli/cloudflare.py
from __future__ import annotations

import argparse
import json

from lib.cloudflare import (
    MANAGE_RULE_SPECS,
    TOKEN_FILE,
    add_rule,
    api_request,
    create_phase_ruleset,
    ensure_local_layout,
    find_phase_ruleset,
    get_ruleset,
    load_config,
    manage_rule_definition,
    masked,
    resolve_zone_id,
    update_rule,
    write_template,
)
from lib.utils.cli import CliError, dispatch


def usage() -> None:
    print(
        """Usage: runseal :cloudflare <command> [args]

Commands:
  init                      create repo-local .local/secrets/cloudflare.env template
  check                     validate repo-local credentials and probe core account APIs
  manage-plan               print the desired manage redirect rule shape
  manage-inspect            inspect current dynamic redirect ruleset for manage rules
  manage-ensure-redirect    create/update exact-path manage redirects (use --dry-run first)
  api <method> <path>       authenticated Cloudflare API call using repo-local token
    [--query key=value]...  optional query params
    [--json <json>]         optional JSON body

Credentials:
  .local/secrets/cloudflare.env
"""
    )


def cmd_init(args: list[str]) -> int:
    if args:
        raise RuntimeError("init does not accept arguments")
    created = write_template()
    ensure_local_layout()
    if created:
        print(f"created {TOKEN_FILE}")
    else:
        print(f"exists {TOKEN_FILE}")
    return 0


def cmd_check(args: list[str]) -> int:
    if args:
        raise RuntimeError("check does not accept arguments")
    config = load_config()
    zone_id = resolve_zone_id(config)
    rulesets = api_request(config, "GET", f"/zones/{zone_id}/rulesets")
    zones = api_request(
        config,
        "GET",
        "/zones",
        params={"account.id": config.account_id, "per_page": "50"},
    )
    account = None
    account_error = False
    try:
        account = api_request(config, "GET", f"/accounts/{config.account_id}")
    except CliError:
        account_error = True
    buckets = None
    buckets_error = False
    try:
        buckets = api_request(config, "GET", f"/accounts/{config.account_id}/r2/buckets")
    except CliError:
        buckets_error = True

    print("cloudflare check: ok")
    print(f"account id: {masked(config.account_id)}")
    if account is not None:
        print(f"account name: {account['result']['name']}")
    elif account_error:
        print("account probe: skipped (token is not authorized for account details)")
    print(f"manage zone: {config.zone_name} ({zone_id})")
    print(f"zone rulesets: {len(rulesets.get('result', []))}")
    print("zones:")
    for zone in zones.get("result", []):
        print(f"  - {zone['name']} ({zone['status']})")
    if buckets is not None:
        print("r2 buckets:")
        for bucket in buckets.get("result", {}).get("buckets", []):
            print(f"  - {bucket['name']}")
    elif buckets_error:
        print("r2 bucket probe: skipped (token is not authorized for R2 bucket list)")
    return 0


def cmd_manage_plan(args: list[str]) -> int:
    if args:
        raise RuntimeError("manage-plan does not accept arguments")
    config = load_config()
    print("manage redirect plan")
    print(f"zone: {config.zone_name}")
    print(f"request host: {config.manage_host}")
    print(f"redirect host: {config.manage_origin_host}")
    print("phase: http_request_dynamic_redirect")
    print("rules:")
    for spec in MANAGE_RULE_SPECS:
        rule = manage_rule_definition(config, spec)
        print(json.dumps(rule, indent=2, sort_keys=True))
    return 0


def cmd_manage_inspect(args: list[str]) -> int:
    if args:
        raise RuntimeError("manage-inspect does not accept arguments")
    config = load_config()
    zone_id = resolve_zone_id(config)
    ruleset = find_phase_ruleset(config, zone_id, phase="http_request_dynamic_redirect")
    if ruleset is None:
        print("manage inspect: no http_request_dynamic_redirect zone ruleset found")
        return 0
    ruleset = get_ruleset(config, zone_id, ruleset["id"])
    print(f"zone id: {zone_id}")
    print(f"ruleset id: {ruleset['id']}")
    print(f"ruleset name: {ruleset['name']}")
    matched = [rule for rule in ruleset.get("rules", []) if rule.get("ref") in {spec.ref for spec in MANAGE_RULE_SPECS}]
    if not matched:
        print("manage inspect: no manage redirect rules found")
        return 0
    print("manage rules:")
    print(json.dumps(matched, indent=2, sort_keys=True))
    return 0


def cmd_manage_ensure_redirect(args: list[str]) -> int:
    parser = argparse.ArgumentParser(prog="runseal :cloudflare manage-ensure-redirect", add_help=False)
    parser.add_argument("--dry-run", action="store_true")
    parsed = parser.parse_args(args)

    config = load_config()
    zone_id = resolve_zone_id(config)
    planned_rules = [manage_rule_definition(config, spec) for spec in MANAGE_RULE_SPECS]

    if parsed.dry_run:
        payload = {
            "zone": config.zone_name,
            "zone_id": zone_id,
            "request_host": config.manage_host,
            "redirect_host": config.manage_origin_host,
            "phase": "http_request_dynamic_redirect",
            "planned_rules": planned_rules,
        }
        print(json.dumps(payload, indent=2, sort_keys=True))
        return 0

    ruleset = find_phase_ruleset(config, zone_id, phase="http_request_dynamic_redirect")
    if ruleset is None:
        ruleset = create_phase_ruleset(
            config,
            zone_id,
            phase="http_request_dynamic_redirect",
            name="Single Redirects ruleset",
        )
    else:
        ruleset = get_ruleset(config, zone_id, ruleset["id"])

    existing = {rule.get("ref"): rule for rule in ruleset.get("rules", [])}
    changed: list[str] = []
    for planned_rule in planned_rules:
        current = existing.get(planned_rule["ref"])
        if current is None:
            add_rule(config, zone_id, ruleset["id"], planned_rule)
            changed.append(f"created {planned_rule['ref']}")
            continue
        update_rule(config, zone_id, ruleset["id"], current["id"], planned_rule)
        changed.append(f"updated {planned_rule['ref']}")

    print("manage ensure redirect: ok")
    for item in changed:
        print(f"  - {item}")
    return 0


def cmd_api(args: list[str]) -> int:
    parser = argparse.ArgumentParser(prog="runseal :cloudflare api", add_help=False)
    parser.add_argument("method")
    parser.add_argument("path")
    parser.add_argument("--query", action="append", default=[])
    parser.add_argument("--json")
    parsed = parser.parse_args(args)

    params: dict[str, str] = {}
    for item in parsed.query:
        if "=" not in item:
            raise RuntimeError(f"invalid --query value: {item}; expected key=value")
        key, value = item.split("=", 1)
        params[key] = value

    body = None
    if parsed.json is not None:
        try:
            body = json.loads(parsed.json)
        except json.JSONDecodeError as err:
            raise RuntimeError(f"invalid --json payload: {err}") from err

    config = load_config()
    payload = api_request(
        config,
        parsed.method,
        parsed.path,
        params=params or None,
        body=body,
    )
    print(json.dumps(payload, indent=2, sort_keys=True))
    return 0


COMMANDS = {
    "init": cmd_init,
    "check": cmd_check,
    "manage-plan": cmd_manage_plan,
    "manage-inspect": cmd_manage_inspect,
    "manage-ensure-redirect": cmd_manage_ensure_redirect,
    "api": cmd_api,
}


def main(argv: list[str] | None = None) -> int:
    return dispatch(argv, usage=usage, commands=COMMANDS, name="cloudflare")


if __name__ == "__main__":
    raise SystemExit(main())

File: .runseal/wrappers/cloudflare.ts
import { env, fail, fileExists, jsonGet, print, runseal, runsealText } from "../lib/runseal.ts";

function usage(): void {
  print("Usage: runseal :cloudflare <command> [args]");
  print("");
  print("Commands:");
  print("  init                      create repo-local .local/secrets/cloudflare.env template");
  print("  check                     validate repo-local credentials and probe core account APIs");
  print("  manage-plan               print the desired manage redirect rule shape");
  print("  manage-inspect            inspect current dynamic redirect ruleset for manage rules");
  print(
    "  manage-ensure-redirect    create/update exact-path manage redirects (use --dry-run first)",
  );
  print(
    "  api                       use: runseal @tool cloudflare api request <method> <path> ...",
  );
  print("");
  print("Credentials:");
  print("  .local/secrets/cloudflare.env");
}

function rejectExtraArg(value: string | undefined, message: string): void {
  if (value !== undefined && value !== "") {
    fail(message);
  }
}

type ManageRules = {
  zoneName: string;
  requestHost: string;
  redirectHost: string;
  ruleSh: string;
  rulePs1: string;
};

async function loadManageRedirectRules(): Promise<ManageRules> {
  const zoneName = await runsealText(["@tool", "cloudflare", "config", "get", "zone_name"]);
  const requestHost = await runsealText(["@tool", "cloudflare", "config", "get", "manage_host"]);
  const redirectHost = await runsealText([
    "@tool",
    "cloudflare",
    "config",
    "get",
    "manage_origin_host",
  ]);
  const prefix = await runsealText([
    "@tool",
    "cloudflare",
    "config",
    "get",
    "manage_redirect_prefix",
  ]);
  const targetSh = prefix === ""
    ? `https://${redirectHost}/manage.sh`
    : `https://${redirectHost}/${prefix}/manage.sh`;
  const targetPs1 = prefix === ""
    ? `https://${redirectHost}/manage.ps1`
    : `https://${redirectHost}/${prefix}/manage.ps1`;
  const ruleSh = await runsealText([
    "@tool",
    "cloudflare",
    "redirect-rule",
    "exact",
    "--ref",
    "runseal_manage_sh_redirect",
    "--description",
    "Redirect runseal manage.sh to releases bucket asset",
    "--host",
    requestHost,
    "--path",
    "/manage.sh",
    "--target-url",
    targetSh,
  ]);
  const rulePs1 = await runsealText([
    "@tool",
    "cloudflare",
    "redirect-rule",
    "exact",
    "--ref",
    "runseal_manage_ps1_redirect",
    "--description",
    "Redirect runseal manage.ps1 to releases bucket asset",
    "--host",
    requestHost,
    "--path",
    "/manage.ps1",
    "--target-url",
    targetPs1,
  ]);
  return { zoneName, requestHost, redirectHost, ruleSh, rulePs1 };
}

async function printManageRedirectPlan(rules: ManageRules, zoneId?: string): Promise<void> {
  const prettySh = await runsealText(["@tool", "json", "pretty", "value", rules.ruleSh]);
  const prettyPs1 = await runsealText(["@tool", "json", "pretty", "value", rules.rulePs1]);
  print("manage redirect plan");
  print(`zone: ${rules.zoneName}`);
  if (zoneId !== undefined) {
    print(`zone id: ${zoneId}`);
  }
  print(`request host: ${rules.requestHost}`);
  print(`redirect host: ${rules.redirectHost}`);
  print("phase: http_request_dynamic_redirect");
  print("rules:");
  print(prettySh);
  print(prettyPs1);
}

const [command, ...rest] = Deno.args;
if (command === undefined || command === "help" || command === "--help") {
  usage();
  Deno.exit(0);
}

switch (command) {
  case "init": {
    rejectExtraArg(rest[0], "cloudflare: init does not accept arguments");
    const localDir = env("RUNSEAL_REPO_LOCAL_DIR", ".local");
    const secretsDir = env("RUNSEAL_REPO_SECRETS_DIR", ".local/secrets");
    const tmpDir = env("RUNSEAL_REPO_TMP_DIR", ".local/tmp");
    const tokenFile = `${secretsDir}/cloudflare.env`;
    await runseal(["@tool", "fs", "mkdir", localDir, "700"]);
    await runseal(["@tool", "fs", "mkdir", secretsDir, "700"]);
    await runseal(["@tool", "fs", "mkdir", tmpDir, "700"]);
    if (await fileExists(tokenFile)) {
      print(`exists ${tokenFile}`);
    } else {
      await runseal([
        "@tool",
        "fs",
        "write-base64",
        tokenFile,
        "IyBSZXBvLWxvY2FsIENsb3VkZmxhcmUgY3JlZGVudGlhbHMgZm9yIHJ1bnNlYWwgc3VwcG9ydCBjb21tYW5kcy4KIyBGaWxsIHRoZXNlIHZhbHVlcyBtYW51YWxseS4gVGhpcyBmaWxlIHN0YXlzIGxvY2FsIGFuZCBnaXRpZ25vcmVkLgpDTE9VREZMQVJFX0FDQ09VTlRfSUQ9CkNMT1VERkxBUkVfQVBJX1RPS0VOPQpDTE9VREZMQVJFX1pPTkVfTkFNRT1wZXJpc2gudWsKQ0xPVURGTEFSRV9NQU5BR0VfSE9TVD1ydW5zZWFsLnBlcmlzaC51awpDTE9VREZMQVJFX01BTkFHRV9PUklHSU5fSE9TVD1yZWxlYXNlcy5ydW5zZWFsLnBlcmlzaC51awpDTE9VREZMQVJFX01BTkFHRV9SRURJUkVDVF9QUkVGSVg9Cg==",
      ]);
      await runseal(["@tool", "fs", "chmod", tokenFile, "600"]);
      print(`created ${tokenFile}`);
    }
    break;
  }
  case "check": {
    rejectExtraArg(rest[0], "cloudflare: check does not accept arguments");
    const accountId = await runsealText(["@tool", "cloudflare", "config", "get", "account_id"]);
    const zoneName = await runsealText(["@tool", "cloudflare", "config", "get", "zone_name"]);
    const zone = await runsealText(["@tool", "cloudflare", "zone", "get", "--name", zoneName]);
    const zoneId = await jsonGet(zone, ".id");
    const rulesets = await runsealText([
      "@tool",
      "cloudflare",
      "zone",
      "ruleset",
      "list",
      "--zone-id",
      zoneId,
    ]);
    const rulesetCount = await runsealText(["@tool", "json", "len", rulesets]);
    const zonesPayload = await runsealText([
      "@tool",
      "cloudflare",
      "api",
      "request",
      "GET",
      "/zones",
      "--query",
      `account.id=${accountId}`,
      "--query",
      "per_page=50",
    ]);
    const zones = await jsonGet(zonesPayload, ".result");
    const zonesPretty = await runsealText(["@tool", "json", "pretty", "value", zones]);
    const account = await runsealText([
      "@tool",
      "cloudflare",
      "account",
      "get",
      "--account-id",
      accountId,
    ]);
    const accountName = await jsonGet(account, ".name");
    const buckets = await runsealText([
      "@tool",
      "cloudflare",
      "account",
      "r2",
      "bucket",
      "list",
      "--account-id",
      accountId,
    ]);
    const bucketsPretty = await runsealText(["@tool", "json", "pretty", "value", buckets]);
    print("cloudflare check: ok");
    print(`account id: ${accountId}`);
    print(`account name: ${accountName}`);
    print(`manage zone: ${zoneName} (${zoneId})`);
    print(`zone rulesets: ${rulesetCount}`);
    print("zones:");
    print(zonesPretty);
    print("r2 buckets:");
    print(bucketsPretty);
    break;
  }
  case "manage-plan": {
    rejectExtraArg(rest[0], "cloudflare: manage-plan does not accept arguments");
    await printManageRedirectPlan(await loadManageRedirectRules());
    break;
  }
  case "manage-inspect": {
    rejectExtraArg(rest[0], "cloudflare: manage-inspect does not accept arguments");
    const zoneName = await runsealText(["@tool", "cloudflare", "config", "get", "zone_name"]);
    const zone = await runsealText(["@tool", "cloudflare", "zone", "get", "--name", zoneName]);
    const zoneId = await jsonGet(zone, ".id");
    const rulesets = await runsealText([
      "@tool",
      "cloudflare",
      "zone",
      "ruleset",
      "list",
      "--zone-id",
      zoneId,
    ]);
    const ruleset = await runsealText([
      "@tool",
      "json",
      "find",
      rulesets,
      "phase",
      "http_request_dynamic_redirect",
    ]);
    if (ruleset === "") {
      print("manage inspect: no http_request_dynamic_redirect zone ruleset found");
      break;
    }
    const rulesetId = await jsonGet(ruleset, ".id");
    const fullRuleset = await runsealText([
      "@tool",
      "cloudflare",
      "zone",
      "ruleset",
      "get",
      "--zone-id",
      zoneId,
      "--ruleset-id",
      rulesetId,
    ]);
    const rulesetName = await jsonGet(fullRuleset, ".name");
    const rules = await jsonGet(fullRuleset, ".rules");
    const matched = await runsealText([
      "@tool",
      "json",
      "filter",
      rules,
      "ref",
      "runseal_manage_sh_redirect",
      "runseal_manage_ps1_redirect",
    ]);
    const matchedCount = await runsealText(["@tool", "json", "len", matched]);
    print(`zone id: ${zoneId}`);
    print(`ruleset id: ${rulesetId}`);
    print(`ruleset name: ${rulesetName}`);
    if (matchedCount === "0") {
      print("manage inspect: no manage redirect rules found");
      break;
    }
    const pretty = await runsealText(["@tool", "json", "pretty", "value", matched]);
    print("manage rules:");
    print(pretty);
    break;
  }
  case "manage-ensure-redirect": {
    let dryRun = false;
    if (rest[0] === "--dry-run") {
      dryRun = true;
      if (rest[1] !== undefined) {
        fail(`cloudflare: unknown manage-ensure-redirect argument: ${rest[1]}`);
      }
    } else if (rest[0] !== undefined) {
      fail(`cloudflare: unknown manage-ensure-redirect argument: ${rest[0]}`);
    }
    const rules = await loadManageRedirectRules();
    const zone = await runsealText([
      "@tool",
      "cloudflare",
      "zone",
      "get",
      "--name",
      rules.zoneName,
    ]);
    const zoneId = await jsonGet(zone, ".id");
    if (dryRun) {
      await printManageRedirectPlan(rules, zoneId);
      break;
    }
    const rulesets = await runsealText([
      "@tool",
      "cloudflare",
      "zone",
      "ruleset",
      "list",
      "--zone-id",
      zoneId,
    ]);
    let ruleset = await runsealText([
      "@tool",
      "json",
      "find",
      rulesets,
      "phase",
      "http_request_dynamic_redirect",
    ]);
    if (ruleset === "") {
      ruleset = await runsealText([
        "@tool",
        "cloudflare",
        "zone",
        "ruleset",
        "create",
        "--zone-id",
        zoneId,
        "--phase",
        "http_request_dynamic_redirect",
        "--name",
        "Single Redirects ruleset",
      ]);
    } else {
      const rulesetId = await jsonGet(ruleset, ".id");
      ruleset = await runsealText([
        "@tool",
        "cloudflare",
        "zone",
        "ruleset",
        "get",
        "--zone-id",
        zoneId,
        "--ruleset-id",
        rulesetId,
      ]);
    }
    const rulesetId = await jsonGet(ruleset, ".id");
    const existingRules = await jsonGet(ruleset, ".rules");
    const currentSh = await runsealText([
      "@tool",
      "json",
      "find",
      existingRules,
      "ref",
      "runseal_manage_sh_redirect",
    ]);
    const currentPs1 = await runsealText([
      "@tool",
      "json",
      "find",
      existingRules,
      "ref",
      "runseal_manage_ps1_redirect",
    ]);
    let changedSh: string;
    if (currentSh === "") {
      await runseal([
        "@tool",
        "cloudflare",
        "zone",
        "ruleset",
        "rule",
        "add",
        "--zone-id",
        zoneId,
        "--ruleset-id",
        rulesetId,
        "--json",
        rules.ruleSh,
      ]);
      changedSh = "created runseal_manage_sh_redirect";
    } else {
      const ruleId = await jsonGet(currentSh, ".id");
      await runseal([
        "@tool",
        "cloudflare",
        "zone",
        "ruleset",
        "rule",
        "update",
        "--zone-id",
        zoneId,
        "--ruleset-id",
        rulesetId,
        "--rule-id",
        ruleId,
        "--json",
        rules.ruleSh,
      ]);
      changedSh = "updated runseal_manage_sh_redirect";
    }
    let changedPs1: string;
    if (currentPs1 === "") {
      await runseal([
        "@tool",
        "cloudflare",
        "zone",
        "ruleset",
        "rule",
        "add",
        "--zone-id",
        zoneId,
        "--ruleset-id",
        rulesetId,
        "--json",
        rules.rulePs1,
      ]);
      changedPs1 = "created runseal_manage_ps1_redirect";
    } else {
      const ruleId = await jsonGet(currentPs1, ".id");
      await runseal([
        "@tool",
        "cloudflare",
        "zone",
        "ruleset",
        "rule",
        "update",
        "--zone-id",
        zoneId,
        "--ruleset-id",
        rulesetId,
        "--rule-id",
        ruleId,
        "--json",
        rules.rulePs1,
      ]);
      changedPs1 = "updated runseal_manage_ps1_redirect";
    }
    print("manage ensure redirect: ok");
    print(`  - ${changedSh}`);
    print(`  - ${changedPs1}`);
    break;
  }
  case "api": {
    if (rest[0] === undefined) {
      fail("cloudflare: api requires a method");
    }
    if (rest[1] === undefined) {
      fail("cloudflare: api requires a path");
    }
    await runseal(["@tool", "cloudflare", "api", "request", ...rest]);
    break;
  }
  default:
    fail(`cloudflare: unknown command: ${command}`);
}
